File: baekjoon/p1516.py
import sys
import heapq
from collections import deque

class Node(object):
	def __init__(self):
		self.targets = list()
		self.self_time = 0
		self.total_time = 0
		self.degree = 0

class MinHeap(object):
	def __init__(self):
		self.h = list()
	
	def push(self, t, e):
		heapq.heappush(self.h, (t, e))
	
	def pop(self):
		return heapq.heappop(self.h)
	
	def __len__(self):
		return len(self.h)


s_num = int(input())

# 1 ~ n
graph = [Node() for _ in range(s_num+1)]
h = MinHeap()

for i in range(1, s_num+1):
	d = 0
	for idx, e in enumerate(map(int, sys.stdin.readline().split())):
		if e == -1:
			continue
		if idx == 0:
			graph[i].self_time = e
			graph[i].total_time = e
		else:
			graph[e].targets.append(i)
			d += 1
	
	graph[i].degree = d
	if d == 0:
		h.push(graph[i].total_time, i)

while len(h) != 0:
	time, cur = h.pop()
	
	for b in graph[cur].targets:
		graph[b].degree -= 1
		if graph[b].degree == 0:
			if time + graph[b].self_time > graph[b].total_time:
				graph[b].total_time = time + graph[b].self_time
				h.push(graph[b].total_time, b)

for i in range(1, len(graph)):
	sys.stdout.write('%d\n' % graph[i].total_time)

File: baekjoon/p1516.cpp
#include <stdio.h>
#include <queue>

#define MAX_S 501

using namespace std;

vector<int> graph[MAX_S];
int self_time[MAX_S] = {0};
int total_time[MAX_S] = {0};

struct Node {
	int time, id;
};

bool operator<(Node a, Node b) {
	return a.time < b.time;
}

priority_queue<Node> q;

int main() {
	int s_num;
	scanf("%d\n", &s_num);

	for (int i = 1; i < s_num+1; i++) {
		int e;
		scanf("%d", &e);
		self_time[i] = e;
		total_time[i] = e;

		bool only_one = true;
		while (true) {
			scanf("%d", &e);
			if (e == -1) 
				break;
			graph[e].push_back(i);
			only_one = false;
		}
		if (only_one) {
			q.push(Node{total_time[i], i});
		}
			
	}
	
	while (!q.empty()) {
		Node n = q.top();
		q.pop();
		int time = n.time;
		int cur = n.id;

		int length = graph[cur].size();
		for (int i = 0; i < length; i++) {
			int branch = graph[cur][i];
			if (time + self_time[branch] > total_time[branch]) {
				total_time[branch] = time + self_time[branch];
				q.push(Node{total_time[branch], branch});
			}
		}
	}

	for (int i = 1; i < s_num+1; i++) {
		printf("%d\n", total_time[i]);
	}

	return 0;
}

